### PR TITLE
Multibyte wb support

### PR DESCRIPTION
**Pull request by jukuisma**
- [x] Mark this if you consider it ready to merge
- [x] I've added tests (optional)
- [x] I wrote some lines in the [book](https://github.com/radareorg/radare2book) (optional): https://github.com/radareorg/radare2-book/pull/403

**Description**

Implement multibyte wb support.

### CODE REVIEW

**Comment by jukuisma** [comment]
TODO:
- [ ] Update pull request description and commit messages
- [ ] Double check the one TODO comment in code and drop the comment
- [ ] Add unit test cases for valid extra spaces and invalid chars
- [ ] Check if the book needs updating

File: libr/core/cmd_write.inc.c [hunk 1]
```diff
@@ -2155,22 +2155,76 @@ static int cmd_wa(void *data, const char *input) {
...
+		}
+		bits_read++;
+
+		// Write a byte if we've read 8 bits
+		if (bits_read % 8 == 0) {
+			r_io_write_at (
+				core->io,
+				// TODO: Do we need some bound check for
+				// `core->offset + block_offset`? Other
+				// functions don't seem to implement any.
```
Comment (jukuisma) [inline libr/core/cmd_write.inc.c:2208]: TODO

File: libr/core/cmd_write.inc.c [hunk 2]
```diff
@@ -2155,22 +2155,76 @@ static int cmd_wa(void *data, const char *input) {
...
+		return 0;
+	}
+
+	// Check that user input only contains binary data
+	for (i = 0; i < uil; i++) {
+		c = input[i];
+		// Ignore whitespaces, \r and \n chars
+		if (c == ' ' || c == '\r' || c == '\n') {
+			continue;
+		}
```
Comment (jukuisma) [inline libr/core/cmd_write.inc.c:2174]: `c == '\r'` is added due to windows. I don't have any windows machines to test on so I'm not sure if this is strictly necessary or not. Shouldn't hurt though.

File: libr/core/cmd_write.inc.c [hunk 3]
```diff
@@ -2155,22 +2155,76 @@ static int cmd_wa(void *data, const char *input) {
...
+	if (uil == 0) {
 		r_core_cmd_help_match (core, help_msg_w, "wb");
+		return 0;
+	}
+
+	// Check that user input only contains binary data
+	for (i = 0; i < uil; i++) {
+		c = input[i];
+		// Ignore whitespaces, \r and \n chars
+		if (c == ' ' || c == '\r' || c == '\n') {
```
Comment (trufae) [inline libr/core/cmd_write.inc.c:2172]: Use isspace() its more portable and fast and comes with libc. Also handles tab

File: libr/core/cmd_write.inc.c [hunk 4]
```diff
@@ -2155,22 +2155,76 @@ static int cmd_wa(void *data, const char *input) {
...
+			continue;
+		}
+		// Check that user input only contains ones and zeros
+		else if (c != '0' && c != '1') {
```
Comment (trufae) [inline libr/core/cmd_write.inc.c:2176]: No need for else after continue

**Review by trufae — CHANGES_REQUESTED** [review]
.

File: libr/core/cmd_write.inc.c (hunk 3 above)
Comment (jukuisma) [reply libr/core/cmd_write.inc.c]: That's way nicer: https://github.com/radareorg/radare2/compare/3a4e307f9574c6c31eb06374ce4587543dceae6c..5e9ade248f5eefe935084f06f334956f3286ab7c 👌

File: libr/core/cmd_write.inc.c (hunk 4 above)
Comment (jukuisma) [reply libr/core/cmd_write.inc.c]: Fixed here: https://github.com/radareorg/radare2/compare/3a4e307f9574c6c31eb06374ce4587543dceae6c..5e9ade248f5eefe935084f06f334956f3286ab7c

File: libr/core/cmd_write.inc.c (hunk 2 above)
Comment (jukuisma) [reply libr/core/cmd_write.inc.c]: `isspace()` handles this for us.

**Comment by trufae** [comment]
Lgtm

**Comment by trufae** [comment]
Still draft?’

**Comment by jukuisma** [comment]
> Still draft?’

Yes, I'll probably add the finishing touches tonight and drop the draft. See: https://github.com/radareorg/radare2/pull/23430#issuecomment-2395980145.

File: libr/core/cmd_write.inc.c (hunk 1 above)
Comment (jukuisma) [reply libr/core/cmd_write.inc.c]: Discussed in discord and tested:
```
╭─vagrant@bookworm ~/scratch/radare2/test  ‹feature-multibyte-wb›
╰─➤  r2 malloc://512
 -- Thanks for using radare2!
[0x00000000]> s 511
[0x000001ff]> wb 11111111 11111111
[0x000001ff]> w0 0x2
[0x000001ff]> w aa
[0x000001ff]>
```
Dropped the TODO comment, should be good. 👌

**Pull request merged**